scripts/components/skillcomponent/skillcomponent.gd

```diff
@@ -0,0 +1,486 @@
+		skills_changed.emit()
+
+
+func remove_skill(skillClass: String):
+	for skill in skills:
+		if skill.skill_class == skillClass:
+			skills.erase(skill)
+			skills_changed.emit()
+			return
+
+
+func skill_select_by_index(index: int):
+	#The index can range from -1 to size()-1
+	if not (index >= -1 and index < skills.size()):
+		GodotLogger.error("Slot index {0} is out of range.".format([str(index)]))  #TEMP
+		return
+
+	#If -1, it is a deselection attempt.
+	if index == -1:
+		skill_deselect()
+		return
+
+	#Otherwise, change the skill properly
+	skill_current = skills[index]
+
+	#This must be separate from the skill_current setter to avoid infinite loops
+	if skill_current.cast_on_select:
+		skill_cast_on_select_selected.emit(skill_current)
+
+
+#Can only find skills inside this component, fails otherwise.
+func skill_select_by_class(skillClass: String):
+	for skill in skills:
+		if skill.skill_class == skillClass:
+			skill_current = skill
+
+			#This must be separate from the skill_current setter to avoid infinite loops
+			if skill_current.cast_on_select:
+				skill_cast_on_select_selected.emit(skill_current)
+
+			return
+
+	GodotLogger.warn('Could not find "{0}" skill in this component'.format([skillClass]))
+
+
+func skill_deselect():
+	skill_current = null
+
+
+func skill_use_at(globalPoint: Vector2, skillClass: String = get_skill_current_class()):
+	var skillUsed: SkillComponentResource = J.skill_resources[skillClass].duplicate()
+	skill_current = skillUsed
+
+	if not is_skill_usable(skillUsed):
+		skill_failed_usage.emit(skillUsed)
+		return
+
+	if user.global_position.distance_to(globalPoint) > skillUsed.hit_range:
+		skill_failed_usage.emit(skillUsed)
+		return
+
+	var skillUsageInfo := UseInfo.new()
+	skillUsageInfo.user = user
+	skillUsageInfo.targets = get_targets(globalPoint)
+	skillUsageInfo.position_target_global = globalPoint
+
+	cooldown_set_time_left(skillUsed.skill_class, skillUsed.cooldown)
+
+	skillUsed.effect(skillUsageInfo)
+	skill_successful_usage.emit(skillUsed)
+
+	if skillUsed.cast_on_select:
+		skill_deselect()
+
+
+func get_targets(where: Vector2) -> Array[Node]:
+	if skill_current.hitbox_rotate_shape:
+		shapeParameters.shape = get_collision_shape(user.global_position.angle_to_point(where))
+
+	else:
+		shapeParameters.shape = get_collision_shape(0.0)
+
+	shapeParameters.collide_with_areas = true
+	shapeParameters.collision_mask = get_collision_layer()
+
+	if not skill_current.hitbox_hits_user:
+		shapeParameters.exclude = [user.get_rid()]
+
+	directSpace = user.get_world_2d().direct_space_state
+	var collisions: Array[Dictionary] = directSpace.intersect_shape(shapeParameters)
+	var targets: Array[Node] = []
+	for coll in collisions:
+		targets.append(coll.get("collider"))
+
+	targets.filter(skill_current._target_filter)
+	return targets
+
+
+func get_collision_shape(userRotation: float) -> Shape2D:
+	var shape: Shape2D
+	#Do not allow shapes with a size of 2
+	assert(skill_current.hitbox_shape.size() != 2)
```

Comment:
Reminder to convert shapes of size() == 2 to Raycast and add methods to generate appropiate points in SkillComponentResource (create_circle_shape(), create_raycast_shape(), create_square_shape(), etc)